feat: Grid constructors with data providers and collection of items

mgarciavaadin opened the pull request:
## Description

It looks convenient to provide constructors with DataProvider/Collection as parameter as it is the common scenario to use Grid with a set of Items defined by a DataProvider or a Collection. 
These new constructors would have also an impact on the modernization of Vaadin 8 to Flow 24 applications as these signatures were implemented in V8 [Grid](https://vaadin.com/api/framework/8.18.0/com/vaadin/ui/Grid.html).

This is the list of the new constructors added:

- public Grid(DataProvider<T,Void> dataProvider)
- public Grid(ListDataProvider dataProvider)
- public Grid(BackEndDataProvider<T, Void> dataProvider)
- public Grid(InMemoryDataProvider dataProvider)
- public Grid(Collection items)

Fixes #4966

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.
- [x] I have not completed some of the steps above and my pull request can be closed immediately.
- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/flow-components?pullRequest=4977) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/flow-components?pullRequest=4977) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/flow-components?pullRequest=4977) it.</sub>

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -1390,6 +1390,80 @@ public Grid() {
+     */
+    public Grid(InMemoryDataProvider<T> inMemoryDataProvider) {
+        this(50);
+        // We don't use DataProvider.withConvertedFilter() here because it's
+        // implementation does not apply the filter converter if Query has a
+        // null filter
+        DataProvider<T, Void> convertedDataProvider = new DataProviderWrapper<T, Void, SerializablePredicate<T>>(
+                inMemoryDataProvider) {
+            @Override
+            protected SerializablePredicate<T> getFilter(Query<T, Void> query) {
+                // Just ignore the query filter (Void) and apply the
+                // predicate only
+                return Optional.ofNullable(inMemoryDataProvider.getFilter())
+                        .orElse(item -> true);
+            }
+        };
+        setDataProvider(convertedDataProvider);
```
It looks like this could reuse the existing `setItems` API, instead of duplicating the code? For consistency I would suggest to do the same for the other new constructors as well.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridDataProviderConstructor.java:
I think a unit test would be enough, it could verify that `Grid.getListDataView`, `Grid.getLazyDataView` return the correct items.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:
```diff
@@ -1390,6 +1390,80 @@ public Grid() {
+     *
+     */
+    public Grid(DataProvider<T, Void> dataProvider) {
+        this(50);
```
No need to duplicate this constant everywhere, we could just reuse the constructor without any arguments, which initializes with a page size of `50` by default.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridDataProviderConstructorTest.java:
```diff
@@ -0,0 +1,354 @@
+        return createColumns(grid);
+    }
+
+    private final Grid<Person> createColumns(Grid<Person> grid) {
```
No need to create columns for this test.

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridDataProviderConstructorTest.java:
```diff
@@ -0,0 +1,354 @@
+        }
+    }
+
+    private final class PersonBackendDataProvider
```
All the custom data providers can extend from `AbstractDataProvider`, which allows removing some of the overrides:
- `refreshItem`
- `refreshItems`
- `refreshAll`
- `addDataProviderListener`

sissbruecker commented on vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridDataProviderConstructorTest.java:
```diff
@@ -0,0 +1,354 @@
+    }
+
+    private final Grid<Person> createGrid(DataProvider dataProvider) {
+        Grid<Person> grid = new Grid<>(dataProvider);
```
The constructor usage here is problematic, because it will effectively only test a single constructor (`public Grid(DataProvider<T, Void> dataProvider)`) for all cases.

Grids should be constructed in their respective test cases with passing the specific type of data provider, rather then the general `DataProvider` type.

sissbruecker commented:
@mgarciavaadin 
- Is there a reason why the PR is still marked as draft?
- Did you address all comments above? If so please add a reply like "Done", or resolve them.
- Please run the formatter

mgarciavaadin commented:
@sissbruecker thanks for your review! Yes I have addressed all your comments and I have run the formatter but the "mvn formatter:format" is not implementing any change although is failing in CI. Locally I just have this diff after running the formatter ```warning: in the working copy of 'vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridDataProviderSizeIT.java', LF will be replaced by CRLF the next time Git touches it``` I'm looking into it

mgarciavaadin commented:
@sissbruecker Thanks for your review! Yes I have addressed your comments and run the formatter. I have moved the request to ready.
